### DISCUSSION TITLE
Clear/overwrite existing signature

### DISCUSSION THREAD

**Discussion by eskuat** [MatthiasValvekens/pyHanko#284] [Q&A] [answered]
Say I have PDF with already filled signature and I want to sign the same field again (or clear then sign) with updated background image for example. How do I do that ?

Steps to reproduce: 

```
pyhanko \
    sign addsig \
    --field form-field-hfejf \
    --style-name sign-style pkcs12 \
    --passfile passfile ./PaperworkPDFS/ppwfile.pdf ./PaperworkPDFS/ppwfile_SIGNED_BY_ESP.pdf ./certif.p12

# Re-add fails: Signature field with name form-field-hfejf appears to be filled already.
pyhanko \
    sign addsig \
    --field form-field-hfejf \
    --style-name sign-style pkcs12 \
    --passfile passfile ./PaperworkPDFS/ppwfile_SIGNED_BY_ESP.pdf ./PaperworkPDFS/ppwfile_SIGNED_BY_ESP.pdf ./certif.p12

```

Second attempt fails with 
`Error: Error raised while producing signed file: Signature field with name form-field-hfejf appears to be filled already.
`

**Comment by MatthiasValvekens** [comment]
This is indeed expected, and not supported in any meaningful sense of the word because there is no universal way to "undo" a signature in a way that preserves the original file's structure, _unless_ you have certain information about how the file was generated. Examples:

  - If you know the signature was generated as an incremental update, reverting to the previous iteration is the most straightforward way to achieve this, but finding the "old" file boundaries programmatically takes some heuristics.
  - If you know there's only one signature in the file, you can regenerate it with the value of the signature field omitted, and start over. This breaks with multiple signers for obvious reasons.
  - ...
  
...which makes me wonder: what's the use case here? Are you sure "overwriting" a signature is really what you want?

**Comment by eskuat** [comment]
I don't have "hard" use case for that, but people tend to make mistakes (applied signature using wrong key for example, or may not like how their hand signature look and want to redo it.) so it would be nice to be able to undo and start everything from the beginning.

I just checked official Adobe Acrobat Reader allows you to clear the signature.

**Reply by MatthiasValvekens** [reply, marked as answer]
> I just checked official Adobe Acrobat Reader allows you to clear the signature.

It might, but I guess that all it does is reverting to the previous revision or something (depending on whether there's a previous signature, maybe). In a CLI context, it's typically just a matter of rerunning the same signing command as before (assuming you kept the original input file around), so I don't feel a pressing need to implement this feature, especially because that minor UX gain comes at a relatively high development/maintenance cost.

I might accept a PR that does this if it's sufficiently robust & maintainable, but barring that I don't see this making it in in the short term. :)
